input/postgres/explain_analyze.go

```diff
@@ -53,16 +58,72 @@ func runExplainAnalyze(ctx context.Context, db *sql.DB, query string, parameters
 }
 
 func validateQuery(query string) error {
-	var isUtil []bool
-	// To be on the safe side never EXPLAIN a statement that can't be parsed,
-	// or multiple statements in one (leading to accidental execution)
-	isUtil, err := util.IsUtilityStmt(query)
-	if err != nil || len(isUtil) != 1 || isUtil[0] {
-		err = fmt.Errorf("query is not permitted to run (multi-statement or utility command?)")
-		return err
+	parseResult, err := pg_query.Parse(query)
+	if err != nil {
+		return fmt.Errorf("query is not permitted to run - failed to parse")
+	}
+	if len(parseResult.Stmts) != 1 {
+		return fmt.Errorf("query is not permitted to run - multi-statement query string")
 	}
 
-	// TODO: Consider adding additional checks here (e.g. blocking known bad function calls)
+	stmt := parseResult.Stmts[0].Stmt.Node
+	switch stmt.(type) {
+	case *pg_query.Node_SelectStmt:
+		// Allowed, continue
+		// Note that we permit wCTEs here (for now), and instead rely on the read-only transaction to block them
+	case *pg_query.Node_InsertStmt, *pg_query.Node_UpdateStmt, *pg_query.Node_DeleteStmt:
+		return fmt.Errorf("query is not permitted to run - DML statement")
+	default:
+		return fmt.Errorf("query is not permitted to run - utility statement")
+	}
+
+	err = validateBlockedFunctions(parseResult)
+	if err != nil {
+		return err
+	}
 
 	return nil
 }
+
+var blockedFunctions = []string{
+	// Blocked because these functions allow exfiltrating data to external servers
+	"dblink",
+	"dblink_connect",
+	"dblink_exec",
+	// Blocked because these functions allow executing arbitrary SQL as input (which can workaround other checks)
+	"xpath_table",
```

Comment:
Is [tablefunc](https://www.postgresql.org/docs/current/tablefunc.html) something we want to block?